stash/projects.go

```diff
@@ -0,0 +1,297 @@
+	groupPermisionsURI = "permissions/groups"
+	userPermisionsURI  = "permissions/users"
+)
+
+// Projects interface defines the methods that can be used to
+// retrieve projects and related permissions.
+type Projects interface {
+	List(ctx context.Context, opts *PagingOptions) (*ProjectsList, error)
+	Get(ctx context.Context, projectName string) (*Project, error)
+	ListProjectGroupsPermission(ctx context.Context, projectKey string, opts *PagingOptions) (*ProjectGroups, error)
+	ListProjectUsersPermission(ctx context.Context, projectKey string, opts *PagingOptions) (*ProjectUsers, error)
+}
+
+// ProjectsService is a client for communicating with stash projects endpoint
+// bitbucket-server API docs: https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html
+type ProjectsService service
+
+// Project represents a Stash project
+// which is a way for teams to group, manage, and organize their repositories.
+type Project struct {
+	// Session is the http.Response of the last request made to the Stash API.
+	Session `json:"sessionInfo,omitempty"`
+	// Description is the project description.
+	Description string `json:"description,omitempty"`
+	// ID is the project ID.
+	ID int64 `json:"id,omitempty"`
+	// Key is the project key.
+	Key string `json:"key,omitempty"`
+	// Links is the project hyperlinks.
+	Links `json:"links,omitempty"`
+	// User is the the authenticated user.
+	User `json:"owner,omitempty"`
+	// Name is the project name.
+	Name string `json:"name,omitempty"`
+	// Public is the project public flag.
+	Public bool `json:"public,omitempty"`
+	// Type is the project type.
+	Type string `json:"type,omitempty"`
+}
+
+// ProjectsList is a list of projects
+type ProjectsList struct {
+	// Paging is the paging information.
+	Paging
+	// Projects is the list of projects.
+	Projects []*Project `json:"values,omitempty"`
+}
+
+// GetProjects returns a slice of Project.
+func (p *ProjectsList) GetProjects() []*Project {
+	return p.Projects
+}
+
+// List retrieves a list of projects.
+// Paging is optional and is enabled by providing a PagingOptions struct.
+// A pointer to a ProjectsList struct is returned. It contains paging information to retrieve the next page of results.
+// List uses the endpoint "GET /rest/api/1.0/projects".
+// bitbucket-server API docs: https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html
+func (s *ProjectsService) List(ctx context.Context, opts *PagingOptions) (*ProjectsList, error) {
+	var query *url.Values
+	query = addPaging(query, opts)
+	req, err := s.Client.NewRequest(ctx, http.MethodGet, newURI(projectsURI), *query, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("get projects request creation failed: %w", err)
+	}
+	res, resp, err := s.Client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("list projects failed: %w", err)
+	}
+
+	// As nothing is done with the response body, it is safe to close here
+	// to avoid leaking connections
+	defer resp.Body.Close()
```

Comment:
You've read the body in the call to `Do`, it should probably close it in there as you're not rewinding?